core/client/src/main/java/alluxio/client/block/AlluxioBlockStore.java

```diff
@@ -40,20 +44,37 @@
 public final class AlluxioBlockStore {
   private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
 
-  private static AlluxioBlockStore sClient = null;
+  // Only one block store will be kept for each master.
+  private static final Map<InetSocketAddress, AlluxioBlockStore> CACHED_BLOCK_STORES =
```

Comment:
We actually need to break up AlluxioBlockStore. If we keep AlluxioBlockStore to be singleton, we need to have a method to change its BlockStoreContext member every time when it gets changed. It seems cleaner to keep the BlockStoreContext member immutable.